### PR TITLE
Smarty5

### PR DESCRIPTION
**Pull request by wisskid**

### CODE REVIEW

**Comment by scottchiefbaker** [comment]
Smarty 4.x was released not too long ago. What is the main goal with Smarty 5?

**Comment by wisskid** [comment]
@scottchiefbaker the codebase has completely been rewritten to use PSR-4, eliminating code duplication and using encapsulation as much as possible, enabling automatic auto-loading using Composer and making the source code much more readable.

**Comment by rudiedirkx** [comment]
I'd go for php 8.0+ then. Only [supported versions](https://www.php.net/supported-versions.php) for this new start. If you're breaking BC, you can break it better. If you want modern code, get a modernish PHP. (8.0 is from 26 Nov 2020)

**Comment by wisskid** [comment]
I've updated the [changelog](https://github.com/smarty-php/smarty/blob/smarty5/CHANGELOG.md) for an overview of changes.

**Comment by lkppo** [comment]
Hello,

I could test this branch.

Will you push it to composer as dev-master/5.0.x-dev?

Regards,

**Comment by wisskid** [comment]
> Hello,
> 
> 
> 
> I could test this branch.
> 
> 
> 
> Will you push it to composer as dev-master/5.0.x-dev?
> 
> 
> 
> Regards,

It turns out you can just use https://packagist.org/packages/smarty/smarty#dev-smarty5

So composer require smarty/smarty:dev-smarty5

**Comment by lkppo** [comment]
> 
> It turns out you can just use https://packagist.org/packages/smarty/smarty#dev-smarty5
> 
> So composer require smarty/smarty:dev-smarty5

Thanks. I will do the test.

**Comment by lkppo** [comment]
> 
> So composer require smarty/smarty:dev-smarty5

I get this error:

```
<br />
<b>Fatal error</b>:  Uncaught Error: Class 'Smarty\Parser\TemplateParser' not found in C:\WinNMP\WWW\wkxlocal\core\libs\vendor\smarty5\smarty\smarty\src\Compiler\Template.php:1230
Stack trace:
#0 C:\WinNMP\WWW\wkxlocal\core\libs\vendor\smarty5\smarty\smarty\src\Compiler\Template.php(425): Smarty\Compiler\Template-&gt;doCompile('&lt;!DOCTYPE html&gt;...', true)
#1 C:\WinNMP\WWW\wkxlocal\core\libs\vendor\smarty5\smarty\smarty\src\Compiler\Template.php(360): Smarty\Compiler\Template-&gt;compileTemplateSource(Object(Smarty\Template))
#2 C:\WinNMP\WWW\wkxlocal\core\libs\vendor\smarty5\smarty\smarty\src\Template\Compiled.php(169): Smarty\Compiler\Template-&gt;compileTemplate(Object(Smarty\Template))
#3 C:\WinNMP\WWW\wkxlocal\core\libs\vendor\smarty5\smarty\smarty\src\Template\Compiled.php(135): Smarty\Template\Compiled-&gt;compileTemplateSource(Object(Smarty\Template))
#4 C:\WinNMP\WWW\wkxlocal\core\libs\vendor\smarty5\smarty\smarty\src\Template\Compiled.php(103): Smarty\Template\Compiled-&gt;compileAndLoad(Object(Smarty\Template))
#5 C:\WinNMP\WWW\wkxloca in <b>C:\WinNMP\WWW\wkxlocal\core\libs\vendor\smarty5\smarty\smarty\src\Compiler\Template.php</b> on line <b>1230</b><br />
```

Should I use the https://github.com/smarty-php/smarty-lexer repo to build a parser?

I'm on Windows. Sould I use flex/bison with ConfigfileParser.y or TemplateParser.y. it's something i never do before.

**Comment by wisskid** [comment]
I've re-added the Lexers and Parsers to VCS, so you should be able to update and run now.

**Comment by lkppo** [comment]
> I've re-added the Lexers and Parsers to VCS, so you should be able to update and run now.

Okay. Thanks.

I tried to use the parser/lexer repo last night installing it by hand and ran into cascading include issues. I try again tonight with your changes.

**Comment by lkppo** [comment]
Do you also include the utillities?

**Comment by lkppo** [comment]
Probably there are a typo in file `C:\WinNMP\WWW\wkxlocal\core\libs\vendor\smarty5\smarty\smarty\src\Smarty.php`, line 857. `Smartyy` should be `Smarty`.

**Comment by lkppo** [comment]
I adapted code from WIKINDX and rewrote a SmartyMenu plugin as an extension for Smarty 5 successfully. Our use of Smarty is not very advanced and I cannot validate everything, but at least I did not encounter any errors.

s Smarty 5 ready or almost ready for production?

**Comment by lkppo** [comment]
When the files are generated in cache, do they take into account the current and previous version number to invalidate the files?

For example to make the transition from Smarty 4 to 5 transparent for the end user?

**Comment by wisskid** [comment]
Yes they do. For each version, actually, so even for patch releases.

**Comment by lkppo** [comment]
I integrated Smarty 5 instead of Smarty 4 in WIKINDX for testing. When everything was well integrated I pushed the code into the VCS and my colleague retrieved it. It had an error that could only be resolved by clearing the build cache so I suspect something is not invalidating the cache properly such as the version.

**Comment by lkppo** [comment]
I also encountered another strange behavior. With Smarty 4.3.0 the use of the array_key_exists() function in a template was not reported as an error whereas with Smarty 5 I needed to register it with registerPlugin('modifier', 'array_key_exists', 'array_key_exists' ). Smarty 4.3.0 should have failed.

**Comment by wisskid** [comment]
4.3 should have reported a deprecation notice, but depending on your error settings, you may or may not have seen this

**Comment by lkppo** [comment]
> 4.3 should have reported a deprecation notice, but depending on your error settings, you may or may not have seen this

I always use `error_reporting = E_ALL` for Smarty, and E_ALL for PHP on my dev machine. The error is:


`Fatal error: Uncaught Error: Call to undefined method Smarty\Template::_decodeProperties() in /Applications/XAMPP/xamppfiles/htdocs/wikindx6/trunk/cache/templates/bryophyta/compile/fab49692e5aec2474eb8f2542098322029bbe8a4_0.file.display.tpl.php:6 Stack trace: #0 /Applications/XAMPP/xamppfiles/htdocs/wikindx6/trunk/core/libs/vendor/smarty/smarty/smarty/src/Template/Compiled.php(263): include() #1 /Applications/XAMPP/xamppfiles/htdocs/wikindx6/trunk/core/libs/vendor/smarty/smarty/smarty/src/Template/Compiled.php(144): Smarty\Template\Compiled->loadCompiledTemplate(Object(Smarty\Template)) #2 /Applications/XAMPP/xamppfiles/htdocs/wikindx6/trunk/core/libs/vendor/smarty/smarty/smarty/src/Template/Compiled.php(108): Smarty\Template\Compiled->compileAndLoad(Object(Smarty\Template)) #3 /Applications/XAMPP/xamppfiles/htdocs/wikindx6/trunk/core/libs/vendor/smarty/smarty/smarty/src/Template.php(184): Smarty\Template\Compiled->render(Object(Smarty\Template)) #4 /Applications/XAMPP/xamppfiles/htdocs/wikindx6/trunk/core/libs/vendor/smarty/smarty/smarty/src/Template.php(639): Smarty\Template->render(false, 1) #5 /Applications/XAMPP/xamppfiles/htdocs/wikindx6/trunk/core/libs/vendor/smarty/smarty/smarty/src/Template.php(576): Smarty\Template->_execute(1) #6 /Applications/XAMPP/xamppfiles/htdocs/wikindx6/trunk/core/libs/vendor/smarty/smarty/smarty/src/Smarty.php(2243): Smarty\Template->display() #7 /Applications/XAMPP/xamppfiles/htdocs/wikindx6/trunk/core/libs/CLOSE.php(234): Smarty\Smarty->display('display.tpl') #8 /Applications/XAMPP/xamppfiles/htdocs/wikindx6/trunk/core/startup/FACTORY.php(465): CLOSE->__construct() #9 /Applications/XAMPP/xamppfiles/htdocs/wikindx6/trunk/index.php(204): FACTORY_CLOSE::getInstance() #10 {main} thrown in /Applications/XAMPP/xamppfiles/htdocs/wikindx6/trunk/cache/templates/bryophyta/compile/fab49692e5aec2474eb8f2542098322029bbe8a4_0.file.display.tpl.phpon line 6`

**Comment by wisskid** [comment]
Ah I didnt think that through. The version check is done in Template::_decodeProperties() but has now moved to another class/method. So no, the compiled templates won't work anymore. I'll make a fix for this.

**Comment by lkppo** [comment]
> Ah I didnt think that through. The version check is done in Template::_decodeProperties() but has now moved to another class/method. So no, the compiled templates won't work anymore. I'll make a fix for this.

Glad the test was useful!

**Comment by marclaporte** [comment]
@wisskid As soon as there is an alpha on https://packagist.org/packages/smarty/smarty, we'll set up a branch to do a daily test upgrade of sites like doc.tiki.org

Details: https://dev.tiki.org/Pre-dogfood-servers-for-Tiki-27-release-process

Thank you very much!

**Comment by wisskid** [comment]
> @wisskid As soon as there is an alpha on https://packagist.org/packages/smarty/smarty, we'll set up a branch to do a daily test upgrade of sites like doc.tiki.org

@marclaporte see https://packagist.org/packages/smarty/smarty#v5.0.0-rc1

Let me know if it behaves

**Comment by deltragon** [comment]
I just wanted to say that this looks amazing, and I am excited for it! Thank you for your hard work!

File: docs/getting-started.md [hunk 1]
```diff
@@ -158,7 +162,7 @@ class Smarty_GuestBook extends Smarty {
 }
 ```
 
-Now, we can use `Smarty_GuestBook` instead of `Smarty` in our scripts:
+Now, we can use `My_GuestBook` instead of `Smarty` in our scripts:
 ```php
 <?php
 $smarty = new Smarty_GuestBook();
```
Comment (davidbuzz) [inline docs/getting-started.md:168]: should be :
$smarty = new My_GuestBook();

**Review by davidbuzz — COMMENTED** [review]
looks ok.

File: docs/getting-started.md (hunk 1 above)
Comment (wisskid) [reply docs/getting-started.md:168]: thanx, fixed it.

**Comment by marclaporte** [comment]
For the record, @rodriguezny @kambereBr and @jacob-js started working on the next steps of the Smarty 5 and Tiki integration.

**Comment by marclaporte** [comment]
We'll continue the updates here: https://github.com/smarty-php/smarty/discussions/920

**Comment by glensc** [comment]
what makes smarty 5? release notes doesn't describe anything...
- https://github.com/smarty-php/smarty/releases/tag/v5.0.0

is there a blog post?

**Comment by marclaporte** [comment]
I am not aware of a blog post, but you can learn more in discussions:
- https://github.com/smarty-php/smarty/discussions/
- https://github.com/smarty-php/smarty/discussions/956

I can tell you it's a huge step forward. Tiki Wiki CMS Groupware uses Tiki since 2002, and we are very happy with Smarty 5. Details:  https://github.com/smarty-php/smarty/discussions/920

Congrats and thank you to @wisskid !

**Pull request merged**